README.adoc

```diff
@@ -213,17 +209,17 @@ Setting this to *true* enables all of the following:
 * `strictPropertyInitialization`: Makes sure you properly initialize class properties _(helpful when extending Foundry's document classes)_
 * `noImplicitThis`: Prevents confusion about what `this` refers to _(particularly important in Foundry's hook callbacks)_
 
-noUncheckedIndexedAccess:: Requires you to verify that array elements or object properties exist before using them.
+noUncheckedIndexedAccess:: Adds `undefined` to types with index signatures. Index signatures are like catch-alls for particular shapes of keys. For example an array in TypeScript will allow expressions like `arr[10]` because that key may exist at runtime. However with `noUncheckedIndexedAccess` set to false (the default) the resulting types will not include `undefined` even though at runtime such a key may not exist.
```

Comment:
I found the original sentence too vague but I honestly struggled to get a terse and still accurate description here. Feedback welcome.

Reply:
Yeah this is a tricky one.  I understand that you have certain information that you want to convey, but I think this change overcorrects and is a bit complex for what I think should be an intuitive, top-level "gist" of what the setting does.  

Language like "the types will include `undefined`" or "adds `undefined` to types..." strikes me as potentially confusing.  It might make perfect sense to someone fluent in TypeScript and type validation in general, but to the layperson it almost sounds like it means the opposite of what it does (i.e. "including `undefined`" could be interpreted as "_permitting_ `undefined`", just as "adds `undefined` to types" almost sounds like you're adding `undefined` to the _acceptable_ types).

How about:

> noUncheckedIndexedAccess:: Requires you to verify that array elements or object properties are defined before using them. By default, TypeScript assumes that any such value that _could_ be defined at runtime _is_ defined, and will not warn you of the possibility that it isn't.  For example, because an array of `string` elements _could_ contain any number of strings at runtime, TypeScript will assume that the type of element `stringArray[632]` _is_ `string`.
> +
> Setting this to *true* prevents TypeScript from making this assumption, by resolving the type of `stringArray[632]` to `string | undefined` instead of `string`.  This requires you to confirm the element really is a `string` before you attempt to use it, preventing runtime errors when accessing dynamic Foundry data like `actor.items[0]` or `token.data.flags["my-module"]`, which might not always exist.

Reply:
> Requires you to verify that array elements or object properties are defined before using them.

I'm happy with the "array elements" part because convention is to call `[x: number, y: number, z: number]` a tuple anyways. However "object properties" is vague and part of what I struggled with. I don't mean to be too nitpicky but I also feel like the wrong direction could be taken with the current language.

Like does `noUncheckedIndexedAccess` mean that `actor.type` has `undefined` added to it? No! That's be so annoying. Does having `noUncheckedIndexedAccess` _off_ mean that `foo: { x?: string }` -> `foo.x` is `string` without `undefined`? No!

It's only for index signatures but explaining that gets tricky.

I overall like the restructuring though.